### ISSUE TITLE
error: no matching function for call to 'std::basic_string<char>::_Alloc_hider::_Alloc_hider(std::basic_string<char>::_Rep*)'

### ISSUE THREAD

**Issue by ilg-ul** [iains/gcc-darwin-arm64#57]
I'm trying to build the _master for the GCC aarch64 development_ from  <https://github.com/iains/gcc-darwin-arm64.git> (actually the default branch, which is `master-wip-apple-si`, commit id 0e0c14134d2d88422e5f39d895011b5e2639d50d), on a Mini M1 running 11.6, with Apple clang, and I get an error while compiling the C++ library.

I ran the offending command separately with `-save-temps` and below is the result:

```console
ilg@xbbma ~ % cd /Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/build/native-gcc-edge-11.2.0/arm64-apple-darwin20.6.0/libstdc++-v3/src/c++11
ilg@xbbma c++11 % /Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/build/native-gcc-edge-11.2.0/./gcc/xgcc -shared-libgcc -B/Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/build/native-gcc-edge-11.2.0/./gcc -nostdinc++ -L/Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/build/native-gcc-edge-11.2.0/arm64-apple-darwin20.6.0/libstdc++-v3/src -L/Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/build/native-gcc-edge-11.2.0/arm64-apple-darwin20.6.0/libstdc++-v3/src/.libs -L/Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/build/native-gcc-edge-11.2.0/arm64-apple-darwin20.6.0/libstdc++-v3/libsupc++/.libs -B/Users/ilg/.local/xbb-edge/usr/arm64-apple-darwin20.6.0/bin/ -B/Users/ilg/.local/xbb-edge/usr/arm64-apple-darwin20.6.0/lib/ -isystem /Users/ilg/.local/xbb-edge/usr/arm64-apple-darwin20.6.0/include -isystem /Users/ilg/.local/xbb-edge/usr/arm64-apple-darwin20.6.0/sys-include -fno-checking -I/Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/sources/gcc-11.2.0-edge/libstdc++-v3/../libgcc -I/Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/build/native-gcc-edge-11.2.0/arm64-apple-darwin20.6.0/libstdc++-v3/include/arm64-apple-darwin20.6.0 -I/Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/build/native-gcc-edge-11.2.0/arm64-apple-darwin20.6.0/libstdc++-v3/include -I/Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/sources/gcc-11.2.0-edge/libstdc++-v3/libsupc++ -std=gnu++11 -D_GLIBCXX_SHARED -fno-implicit-templates -Wall -Wextra -Wwrite-strings -Wcast-qual -Wabi=2 -fdiagnostics-show-location=once -ffunction-sections -fdata-sections -frandom-seed=cow-stdexcept.lo -g -pipe -O2 -ffunction-sections -fdata-sections -mmacosx-version-min=11.0 -w -c /Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/sources/gcc-11.2.0-edge/libstdc++-v3/src/c++11/cow-stdexcept.cc  -fno-common -DPIC -D_GLIBCXX_SHARED -o cow-stdexcept.o -save-temps
xgcc: warning: '-pipe' ignored because '-save-temps' specified
In file included from /Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/build/native-gcc-edge-11.2.0/arm64-apple-darwin20.6.0/libstdc++-v3/include/bits/basic_string.h:52,
                 from /Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/build/native-gcc-edge-11.2.0/arm64-apple-darwin20.6.0/libstdc++-v3/include/string:53,
                 from /Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/build/native-gcc-edge-11.2.0/arm64-apple-darwin20.6.0/libstdc++-v3/include/stdexcept:39,
                 from /Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/sources/gcc-11.2.0-edge/libstdc++-v3/src/c++11/cow-stdexcept.cc:48:
/Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/sources/gcc-11.2.0-edge/libstdc++-v3/include/bits/cow_string.h: In instantiation of 'std::basic_string<_CharT, _Traits, _Alloc>::basic_string(std::basic_string<_CharT, _Traits, _Alloc>&&) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]':
/Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/build/native-gcc-edge-11.2.0/arm64-apple-darwin20.6.0/libstdc++-v3/include/bits/basic_string.h:4019:12:   required from here
/Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/sources/gcc-11.2.0-edge/libstdc++-v3/include/bits/cow_string.h:630:9: error: no matching function for call to 'std::basic_string<char>::_Alloc_hider::_Alloc_hider(std::basic_string<char>::_Rep*)'
  630 |       : _M_dataplus(__str._M_rep())
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/build/native-gcc-edge-11.2.0/arm64-apple-darwin20.6.0/libstdc++-v3/include/bits/basic_string.h:52,
                 from /Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/build/native-gcc-edge-11.2.0/arm64-apple-darwin20.6.0/libstdc++-v3/include/string:53,
                 from /Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/build/native-gcc-edge-11.2.0/arm64-apple-darwin20.6.0/libstdc++-v3/include/stdexcept:39,
                 from /Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/sources/gcc-11.2.0-edge/libstdc++-v3/src/c++11/cow-stdexcept.cc:48:
/Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/sources/gcc-11.2.0-edge/libstdc++-v3/include/bits/cow_string.h:315:9: note: candidate: 'std::basic_string<_CharT, _Traits, _Alloc>::_Alloc_hider::_Alloc_hider(_CharT*, const _Alloc&) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]'
  315 |         _Alloc_hider(_CharT* __dat, const _Alloc& __a) _GLIBCXX_NOEXCEPT
      |         ^~~~~~~~~~~~
/Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/sources/gcc-11.2.0-edge/libstdc++-v3/include/bits/cow_string.h:315:9: note:   candidate expects 2 arguments, 1 provided
/Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/sources/gcc-11.2.0-edge/libstdc++-v3/include/bits/cow_string.h:313:14: note: candidate: 'std::basic_string<char>::_Alloc_hider::_Alloc_hider(const std::basic_string<char>::_Alloc_hider&)'
  313 |       struct _Alloc_hider : _Alloc
      |              ^~~~~~~~~~~~
/Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/sources/gcc-11.2.0-edge/libstdc++-v3/include/bits/cow_string.h:313:14: note:   no known conversion for argument 1 from 'std::basic_string<char>::_Rep*' to 'const std::basic_string<char>::_Alloc_hider&'
/Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/sources/gcc-11.2.0-edge/libstdc++-v3/include/bits/cow_string.h:313:14: note: candidate: 'std::basic_string<char>::_Alloc_hider::_Alloc_hider(std::basic_string<char>::_Alloc_hider&&)'
/Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/sources/gcc-11.2.0-edge/libstdc++-v3/include/bits/cow_string.h:313:14: note:   no known conversion for argument 1 from 'std::basic_string<char>::_Rep*' to 'std::basic_string<char>::_Alloc_hider&&'
In file included from /Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/build/native-gcc-edge-11.2.0/arm64-apple-darwin20.6.0/libstdc++-v3/include/bits/basic_string.h:52,
                 from /Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/build/native-gcc-edge-11.2.0/arm64-apple-darwin20.6.0/libstdc++-v3/include/string:53,
                 from /Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/build/native-gcc-edge-11.2.0/arm64-apple-darwin20.6.0/libstdc++-v3/include/stdexcept:39,
                 from /Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/sources/gcc-11.2.0-edge/libstdc++-v3/src/c++11/cow-stdexcept.cc:48:
/Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/sources/gcc-11.2.0-edge/libstdc++-v3/include/bits/cow_string.h: In instantiation of 'std::basic_string<_CharT, _Traits, _Alloc>::basic_string(std::basic_string<_CharT, _Traits, _Alloc>&&) [with _CharT = wchar_t; _Traits = std::char_traits<wchar_t>; _Alloc = std::allocator<wchar_t>]':
/Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/build/native-gcc-edge-11.2.0/arm64-apple-darwin20.6.0/libstdc++-v3/include/bits/basic_string.h:4152:44:   required from here
/Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/sources/gcc-11.2.0-edge/libstdc++-v3/include/bits/cow_string.h:630:9: error: no matching function for call to 'std::basic_string<wchar_t>::_Alloc_hider::_Alloc_hider(std::basic_string<wchar_t>::_Rep*)'
  630 |       : _M_dataplus(__str._M_rep())
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/build/native-gcc-edge-11.2.0/arm64-apple-darwin20.6.0/libstdc++-v3/include/bits/basic_string.h:52,
                 from /Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/build/native-gcc-edge-11.2.0/arm64-apple-darwin20.6.0/libstdc++-v3/include/string:53,
                 from /Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/build/native-gcc-edge-11.2.0/arm64-apple-darwin20.6.0/libstdc++-v3/include/stdexcept:39,
                 from /Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/sources/gcc-11.2.0-edge/libstdc++-v3/src/c++11/cow-stdexcept.cc:48:
/Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/sources/gcc-11.2.0-edge/libstdc++-v3/include/bits/cow_string.h:315:9: note: candidate: 'std::basic_string<_CharT, _Traits, _Alloc>::_Alloc_hider::_Alloc_hider(_CharT*, const _Alloc&) [with _CharT = wchar_t; _Traits = std::char_traits<wchar_t>; _Alloc = std::allocator<wchar_t>]'
  315 |         _Alloc_hider(_CharT* __dat, const _Alloc& __a) _GLIBCXX_NOEXCEPT
      |         ^~~~~~~~~~~~
/Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/sources/gcc-11.2.0-edge/libstdc++-v3/include/bits/cow_string.h:315:9: note:   candidate expects 2 arguments, 1 provided
/Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/sources/gcc-11.2.0-edge/libstdc++-v3/include/bits/cow_string.h:313:14: note: candidate: 'std::basic_string<wchar_t>::_Alloc_hider::_Alloc_hider(const std::basic_string<wchar_t>::_Alloc_hider&)'
  313 |       struct _Alloc_hider : _Alloc
      |              ^~~~~~~~~~~~
/Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/sources/gcc-11.2.0-edge/libstdc++-v3/include/bits/cow_string.h:313:14: note:   no known conversion for argument 1 from 'std::basic_string<wchar_t>::_Rep*' to 'const std::basic_string<wchar_t>::_Alloc_hider&'
/Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/sources/gcc-11.2.0-edge/libstdc++-v3/include/bits/cow_string.h:313:14: note: candidate: 'std::basic_string<wchar_t>::_Alloc_hider::_Alloc_hider(std::basic_string<wchar_t>::_Alloc_hider&&)'
/Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/sources/gcc-11.2.0-edge/libstdc++-v3/include/bits/cow_string.h:313:14: note:   no known conversion for argument 1 from 'std::basic_string<wchar_t>::_Rep*' to 'std::basic_string<wchar_t>::_Alloc_hider&&'
ilg@xbbma c++11 % 
```

The two temporary files are:

[cow-stdexcept.zip](https://github.com/iains/gcc-darwin-arm64/files/7617353/cow-stdexcept.zip)

The script is the same as when building my development GCC, just that is uses a different build folder, and the source is obtained with `git clone` instead of downloading an archive.

**Comment by iains** [comment]
When reporting a build error we need;
- the configure line.
- what the bootstrapping compiler is (you mention clang, but not which version)

Works for Me™

$ clang --version
Apple clang version 12.0.5 (clang-1205.0.22.9)
Target: arm64-apple-darwin20.4.0

$ ..../src-local/gcc-master/configure --prefix=/Users/iains/aarch64-apple-darwin20/gcc-12-0-0 --build=aarch64-apple-darwin20 --with-sysroot=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk --enable-languages=all CC=clang CXX=clang++ --enable-pie-tools  >conf.txt
configure: WARNING: using in-tree isl, disabling version check
configure: WARNING: GNAT is required to build ada
configure: WARNING: go not supported for this target
*** This configuration is not supported in the following subdirectories:
     target-libsanitizer target-libvtv target-libphobos target-libffi target-libgo gnattools gotools target-libada target-zlib target-liboffloadmic
...
$ time nice make -j8 >b.txt 2>e.txt ....

real	55m45.483s
user	239m10.610s
sys	16m24.194s

**Comment by ilg-ul** [comment]
The compiler is the one installed via Xcode/CLT:

```console
% clang --version
Apple clang version 13.0.0 (clang-1300.0.29.3)
Target: arm64-apple-darwin20.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
% uname -a
Darwin xbbma.local 20.6.0 Darwin Kernel Version 20.6.0: Tue Oct 12 18:33:38 PDT 2021; root:xnu-7195.141.8~1/RELEASE_ARM64_T8101 arm64
```

The configure line is a bit more elaborate than yours:

```console
$ bash /Users/ilg/Work/xbb-edge-3.4-macos-11.6.1-arm64/sources/gcc-11.2.0-edge/configure --prefix=/Users/ilg/.local/xbb-edge/usr --libdir=/Users/ilg/.local/xbb-edge/usr/arm64-apple-darwin20.6.0/lib --program-suffix=-xbb --build=arm64-apple-darwin20.6.0 --target=arm64-apple-darwin20.6.0 --with-pkgversion=xPack Build Box arm64 GCC --with-dwarf2 --with-stabs --with-libiconv --with-isl=/Users/ilg/.local/xbb-edge --with-gnu-as --with-gnu-ld --with-diagnostics-color=auto --without-system-zlib --without-cuda-driver --enable-languages=c,c++,fortran,objc,obj-c++,lto --enable-objc-gc=auto --enable-checking=release --enable-lto --enable-plugin --enable-static --enable-__cxa_atexit --enable-threads=posix --enable-gnu-unique-object --enable-gnu-indirect-function --enable-linker-build-id --enable-fully-dynamic-string --enable-cloog-backend=isl --enable-libgomp --enable-libssp --enable-default-pie --enable-default-ssp --enable-libatomic --enable-graphite --enable-libquadmath --enable-libquadmath-support --enable-libstdcxx --enable-libstdcxx-time=yes --enable-libstdcxx-visibility --enable-libstdcxx-threads --with-default-libstdcxx-abi=new --enable-bootstrap --disable-nls --disable-multilib --disable-libstdcxx-debug --disable-libstdcxx-pch --disable-install-libiberty --disable-werror --enable-shared --enable-shared-libgcc --with-sysroot=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk
```

Exactly the same configuration works fine with the HomeBrew branch and the upstream GCC.

Could you check if the offending file was actually compiled in your build? (`libstdc++-v3/src/c++11/cow-stdexcept.cc`)

**Comment by iains** [comment]
1. I cannot change the version of  Xcode on gcc304 (it's not my machine).

2. that's an amazing configure line - you should not need to add any configure options other than telling it where the sysroot is and where you want the paths to be (plus the one or two that relate to items that are WIP).

the following options are definitely wrong:
--with-stabs  (not supported by Xcode assemblers for a long time)
--with-gnu-as (no, we use the ```as``` from Xcode .. unless you've made a port of GAS?)
--with-gnu-ld (no, we use Apple's ```ld64``` .. unless you've made a port of GNU ld?)
 
many of the other config options are either irrelevant to Darwin (and thus it is not tested as to whether they do something bad) or are just repeating defaults.

The golden rule with a configuration option is **"do not add it unless you know exactly why you need it".**

**Comment by ilg-ul** [comment]
> that's an amazing configure line ...

Yes, it is probably excessively long. It ended up like this because the same scripts also builds similar tools on Intel 32/64 and Arm 32/64 Linux. 

And exactly the same configuration worked with GNU GCC up to 11.2. I would like to avoid a future 11.3 that fails to build. 

But can you confirm that the `libstdc++-v3/src/c++11/cow-stdexcept.cc` file was actually compiled in your build, and the command line is similar?

**Comment by iains** [comment]
- The exact purpose of configure scripts is to choose the correct values for each platform (certainly with some customisation for placement of the output and which languages you might want to build - but as said, only add those you know you need).

- overriding the choices the script would make defeats the very thing you want;  that the same sources work on multiple platforms.

- The fact that those config lines "work" in other cases could well be pure luck - the values I mentioned above are not valid for x86 Darwin either.

- my build completed sucessfully, so yes that object was built.

- I will set an x86-64 build going with clang-13 (in case that is somehow miscompiling stage #1, but that is somewhat of a long shot).

**Comment by ilg-ul** [comment]
> so yes that object was built.

Could you do a `grep cow-stdexcept.cc b.txt`, and copy here the compile line of that file, so I can try to identify what is different in my build?

**Comment by ilg-ul** [comment]
> I cannot change the version of Xcode on gcc304 (it's not my machine).

If you have a page explaining the prerequisites and how to configure and run the builds, I can try exactly the same on my machine.

**Comment by iains** [comment]
>>so yes that object was built.
>Could you do a grep cow-stdexcept.cc b.txt, and copy here the compile line of that file, so I can try to identify what is different in my build?

space is tight on gcc304 - the builds are already deleted (but, JFTR your command line doesn't look like a bootstrap internal command - the compiler seems to be GCC not clang?)

>>I cannot change the version of Xcode on gcc304 (it's not my machine).
>If you have a page explaining the prerequisites and how to configure and run the builds, I can try exactly the same on my machine.

I showed you my configure line above (that uses the Xcode CLT install; you should not need more than that for building - for testing, of course, you have to install dejagnu).

**Comment by ilg-ul** [comment]
>  the compiler seems to be GCC not clang

Oops! You are right, the build used my gcc bootstrap, which is based on HomeBrew gcc-11.2.0-arm-20211126, not the Apple clang, as I intended.

I'll redo the test.

**Comment by ilg-ul** [comment]
I did some further testing and the culprit is `--enable-fully-dynamic-string`.

The error does not depend on the bootstrap, it occurs with both clang and GCC 11.2.

I also did some cleanups in the configure options, but this was not relevant for the current issue.

Is there something special with `--enable-fully-dynamic-string`? 

I'll disable it for now, and proceed with further tests. After @fxcoudert will merge the commit fixing #55 I'll update again the development environment to the latest HomeBrew branch and redo all application builds.

**Comment by iains** [comment]
```--enable-fully-dynamic-string
                          do not put empty strings in per-process static
                          memory [default=no]```

So why have you overridden the default?
what problem is it solving? 

perhaps that is a real bug and the configure option was just hiding the problem?

**Comment by ilg-ul** [comment]
I have difficulties to follow you.

The option was there because in one of the previous releases (the scripts were maintained since macOS 10.10), it was required to match the settings of the macOS system libraries. 

If for recent systems it is no longer needed, I'll remove it, but, remember, the same build scripts should run on both macOS 11.6 (the base for Apple Silicon systems) and macOS 10.13 (the base for Intel systems).

**Comment by iains** [comment]
> I have difficulties to follow you.

likewise,..

> The option was there because in one of the previous releases (the scripts were maintained since macOS 10.10), it was required to match the settings of the macOS system libraries.

The system libraries have not be built with GCC since 10.7

How have you determined "match the settings of the macOS system libraries"?

Which test cases pass with this option enabled that fail without  it?

If we have a bug with string handling we should fix it.
Iff the option is needed on macOS to be compatible then we should default it on.

The point is that configure options are supposed to have the correct defaults for targets.

**Comment by iains** [comment]
please could you show the output of:

```
nm -mg <target>/libstdc++-v3/src/.libs/libstdc++.6.dylib |grep _ZNSs4_Rep20_S_empty_rep_storageE
```
where  <target>  is x86_64-apple-darwinMM or aarach64-apple-darwinNN

**Comment by fxcoudert** [comment]
As far as I can see, Homebrew never used `--enable-fully-dynamic-string` (https://github.com/Homebrew/legacy-homebrew/issues/23224). MacPorts stopped using it 10 years ago: https://trac.macports.org/ticket/27237 and it was already problematic.

I think this was (maybe) necessary when Apple's system compiler was GCC. The last such version was shipped as part of Xcode 4.1, in 2011, for macOS 10.6.

**Comment by iains** [comment]
I was discussing with the libstdc++ main developer - we shouldn't need this option on Darwin, since the library exports a weak version.

However there's still the horrible problem with the installed /usr/lib/libstdc++ having duplicate symbols - if someone screws up and links that one into the same process as the FSF current library.  This should be much less of an issue on macOS 10.8+ since AFAIR there are no system uses of libstdc++ from then (there are very few in 10.7 IIRC)

**Comment by ilg-ul** [comment]
>  show the output of:

On a freshly build from the default branch of <https://github.com/iains/gcc-darwin-arm64.git> with Apple clang on a Mini M1 macOS 11.6, with minimal configuration options:

```console
ilg@xbbma native-gcc-edge-11.2.0 % nm -mg arm64-apple-darwin20.6.0/libstdc++-v3/src/.libs/libstdc++.6.dylib |grep _ZNSs4_Rep20_S_empty_rep_storageE 
0000000000193428 (__DATA,__data) weak external __ZNSs4_Rep20_S_empty_rep_storageE
```

**Comment by iains** [comment]
right, that's correct (and the same as I get) that means that there will only be one instance of  __ZNSs4_Rep20_S_empty_rep_storageE  (std::string::_Rep::_S_empty_rep_storage) in a process (which is what is required) - having some things built with  ```--enable-fully-dynamic-string``` means that they have a different ABI from things built without (and will break at runtime with segvs),.

**Comment by ilg-ul** [comment]
Let me summarise:

- you suggested to report Apple Silicon GCC bugs only after testing to the latest master in this repo, which I had not at hand
- to build it, I reused an older script, which is definitely not ideal, but previously passed the build for the Intel GNU 11.2, so I considered a good starting point
- the build failed in libstdc++ with that error message
- based on your suggestions, I simplified the configuration options, and concluded that the error was caused by the presence of `--enable-fully-dynamic-string` 
- I do not have a good reason for using this option on macOS 11.6 arm64, it was there for historical reasons, and I don't mind removing it, as long as the Intel builds on macOS 10.13 will not be affected
- I do not know if this option should be enabled by default or not, but, for my use case, I think that it is mandatory for the resulting object files to be compatible with those created by the system Apple clang, both on Apple Silicon macOS 11.6 and Intel macOS 10.13

In my scripts I already removed this option, so I'm fine, we can close this ticket.

But in general, if the presence of this option is harmful for Apple Silicon GCC, or for macOS GCC in general, I think that it would be preferable for the configure script to complain, and do not start the build, to later fail during building libstdc++.

Of course that there could be better strategies that I am not aware of, so feel free to take any actions you think necessary, including ignoring this issue and closing the ticket.

**Comment by ilg-ul** [comment]
BTW, this hopefully concludes porting my development environment (xPack Build Box) to Apple SIlicon the remaining issue that affect several packages that can be compiled only with clang are cause by the use of `static const` as an array size, feature not yet supported by GCC.

I'll try to freeze the XBB scripts for now, and update them when 11.3 will be out.

If you think useful, I can do a test run when you'll have a pre-release of 11.3, to confirm that everything is ok.

**Comment by ilg-ul** [comment]
I confirm that with the latest patches, included in the HomeBrew [gcc-11.2.0-arm-20211201](https://github.com/fxcoudert/gcc/releases/tag/gcc-11.2.0-arm-20211201), I was able to successfully build the xPack Build Box development environment on a Mac Mini M1 running macOS 11.6.1.

This is not an easy feat, since XBB is a relatively large set of tools, (almost) all compiled with the new GCC 11.2:

```
ilg@xbbma xbb % ls  bin usr/bin 
bin:
2to3			git			mktemp			rst2xml.py
2to3-3.9		git-cvsserver		mpicalc			rstpep2html.py
7za			git-receive-pack	msgattrib		runcon
7zr			git-shell		msgcat			scons
[			git-upload-archive	msgcmp			scons-configure-cache
__pycache__		git-upload-pack		msgcomm			sconsign
aclocal			gitk			msgconv			sdiff
aclocal-1.16		glibtool		msgen			sed
asn1Coding		glibtoolize		msgexec			seq
asn1Decoding		gm4			msgfilter		sfv-hash
asn1Parser		gmake			msgfmt			sha1sum
autoconf		gnutar			msggrep			sha224sum
autogen			gnutls-cli		msginit			sha256sum
autoheader		gnutls-cli-debug	msgmerge		sha384sum
autom4te		gnutls-serv		msgunfmt		sha512sum
automake		gost12-256-hash		msguniq			shasum
automake-1.16		gost12-512-hash		mv			shred
autoopts-config		gpg			ncurses6-config		shuf
autopoint		gpg-agent		ngettext		sleep
autoreconf		gpg-card		nice			sort
autoscan		gpg-connect-agent	ninja			sphinx-apidoc
autoupdate		gpg-error		nl			sphinx-autogen
awk			gpg-error-config	nohup			sphinx-build
b2sum			gpg-wks-client		normalizer		sphinx-quickstart
base32			gpg-wks-server		nproc			splain
base64			gpgconf			npth-config		split
basename		gpgparsemail		numfmt			sqlite3_analyzer
basenc			gpgrt-config		ocsptool		srptool
bash			gpgscm			od			stat
bashbug			gpgsm			openssl			stdbuf
bison			gpgsplit		paste			streamzip
c_rehash		gpgtar			patch			stty
captoinfo		gpgv			patchelf		sum
cat			groups			pathchk			sync
ccmake			gsed			pdftexi2dvi		tabs
certtool		guild			perl			tac
chcon			guile			perl5.34.0		tail
chgrp			guile-config		perlbug			tar
chmod			guile-snarf		perldoc			tclsh8.6
chown			guile-tools		perlivp			tee
chroot			h2ph			perlthanks		test
cksum			h2xs			piconv			texi2any
clear			has160-hash		pinky			texi2dvi
cmake			head			pip			texi2pdf
cmp			hmac256			pip3			texindex
columns			hostid			pip3.9			tic
comm			id			pkg-config		tiger-hash
corelist		idle3			pkg-config-verbose	timeout
cp			idle3.9			pl2pm			toe
cpack			ifnames			png-fix-itxt		touch
cpan			info			pngfix			tput
cpanm			infocmp			pod2html		tr
csplit			infotocap		pod2man			true
ctest			install			pod2texi		truncate
curl			install-info		pod2text		tset
curl-config		instmodsh		pod2usage		tsort
cut			join			podchecker		tth-hash
date			json_pp			pr			tty
dd			kbxutil			printenv		uname
df			kill			printf			unexpand
diff			ksba-config		prove			uniq
diff3			libassuan-config	psktool			unix2dos
dir			libgcrypt-config	ptar			unix2mac
dircolors		libnetcfg		ptardiff		unlink
dirmngr			libpng-config		ptargrep		unlzma
dirmngr-client		libpng16-config		ptx			unxz
dirname			libtool			pwd			uptime
dos2unix		libtoolize		pybabel			users
du			link			pydoc3			vdir
dumpsexp		ln			pydoc3.9		watchgnupg
easy_install		logname			pygmentize		wc
easy_install-3.8	ls			python3			wget
easy_install-3.9	lzcat			python3-config		wheel
echo			lzcmp			python3.9		whirlpool-hash
ed2k-link		lzdiff			python3.9-config	who
edonr256-hash		lzegrep			re2c			whoami
edonr512-hash		lzfgrep			re2go			xgettext
enc2xs			lzgrep			readlink		xml2ag
encguess		lzless			realpath		xmlwf
env			lzma			recode-sr-latin		xsubpp
envsubst		lzmadec			reset			xz
expand			lzmainfo		rhash			xzcat
expr			lzmore			rm			xzcmp
factor			m4			rmdir			xzdec
false			mac2unix		rst2html.py		xzdiff
fmt			magnet-link		rst2html4.py		xzegrep
fold			make			rst2html5.py		xzfgrep
gawk			makedepend		rst2latex.py		xzgrep
gawk-5.1.1		makeinfo		rst2man.py		xzless
get-gcc-rpath		md5sum			rst2odt.py		xzmore
getdefs			meson			rst2odt_prepstyles.py	yacc
gettext			mkdir			rst2pseudoxml.py	yat2m
gettext.sh		mkfifo			rst2s5.py		yes
gettextize		mknod			rst2xetex.py		zipdetails

usr/bin:
arm64-apple-darwin20.6.0-c++-xbb	gcc-ar
arm64-apple-darwin20.6.0-g++-xbb	gcc-ar-xbb
arm64-apple-darwin20.6.0-gcc-11.2.0	gcc-nm
arm64-apple-darwin20.6.0-gcc-ar-xbb	gcc-nm-xbb
arm64-apple-darwin20.6.0-gcc-nm-xbb	gcc-ranlib
arm64-apple-darwin20.6.0-gcc-ranlib-xbb	gcc-ranlib-xbb
arm64-apple-darwin20.6.0-gcc-xbb	gcc-xbb
arm64-apple-darwin20.6.0-gfortran-xbb	gcov
c++					gcov-dump
c++-xbb					gcov-dump-xbb
cc					gcov-tool
cpp-xbb					gcov-tool-xbb
g++					gcov-xbb
g++-xbb					gfortran-xbb
gcc					lto-dump-xbb
ilg@xbbma xbb % 
```

Of this list, only `pkg_config`, `cmake` and `python` required `clang`, since they include a system header that use a `static const` as an array size, feature not yet supported by GCC.

The same script also successfully built XBB on an Intel Mac running macOS 10.13, but using the official GNU GCC 11.2 sources.

The next step is to use this development environment to re-build the xPack cross-platform standalone binaries, like gcc, clang, arm-none-eabi-gcc, riscv-none-embed-gcc, cmake, ninja, meson, etc, but I don't expect many problems with the compiler itself.

**Comment by fxcoudert** [comment]
> Of this list, only pkg_config, cmake and python required clang, since they include a system header that use a static const as an array size, feature not yet supported by GCC.

Asking for the record: is the header `Authorization.h` (and its `kAuthorizationExternalFormLength` constant)? I filed a bug against it in 2017 (FB5441805), because a system header should not include nonstandard C code… but Apple has not followed up.

**Comment by iains** [comment]
> Of this list, only `pkg_config`, `cmake` and `python` required `clang`, since they include a system header that use a `static const` as an array size, feature not yet supported by GCC.

For the record, this is because the SDK headers are using a non-standard clang extension to C (those values are not constants in standard C, only in C++).

I have a workaround patch for that - but do not know if it will 
be acceptable to upstream GCC (however, the historically-released SDKs will not be changed, and fixincludes does not work on Frameworks - so I think that the patch will be a 'vendor branch' candidate)

**Comment by ilg-ul** [comment]
>  is the header `Authorization.h`

Thar's correct, it is the `/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/Security.framework/Versions/A/Headers/Authorization.h` header and the incompatible statements are:

```
static const size_t kAuthorizationExternalFormLength = 32;
/*!
        @struct AuthorizationExternalForm
        An AuthorizationExternalForm structure can hold the externalized form of
        an AuthorizationRef. As such, it can be transmitted across IPC channels
        to other processes, which can re-internalize it to recover a valid AuthorizationRef
        handle.
        The data contained in an AuthorizationExternalForm should be considered opaque.

        SECURITY NOTE: Applications should take care to not disclose the AuthorizationExternalForm to
        potential attackers since it would authorize rights to them.
*/
typedef struct {
        char bytes[kAuthorizationExternalFormLength];
} AuthorizationExternalForm;
```

**Comment by iains** [comment]
> Asking for the record: is the header `Authorization.h` (and its `kAuthorizationExternalFormLength` constant)? I filed a bug against it in 2017 (FB5441805), because a system header should not include nonstandard C code… but Apple has not followed up.

I believe that there are several headers with this issue - one also in one of the graphics frameworks IIRC (I think I had to fix this to be able to build cmake 16.1 but maybe that's poor memory)

**Comment by ilg-ul** [comment]
BTW, I think you should enable Discussion for this project, to avoid polluting the issues with side discussions.

**Comment by iains** [comment]
In any case, Apple cannot undo this in released SDKs and I would not expect them to re-release the SDKs for unsupported OS version.

**Comment by ilg-ul** [comment]
> I would not expect them to re-release the SDKs

I would not expect them either. 

The solution would be to either support this in GCC. The workaround would be to avoid including those offending headers.

**Comment by iains** [comment]
> The solution would be to support this in GCC.

* As noted, I cannot be certain that such a work-around will be acceptable to upstream GCC (but it will be presented when I clear the more important patch queue)

* We can do this definitely in  a "Darwin vendor" branch (it is quite conventional for distributions to have patches that are not present 'upstream')

> The workaround would be to avoid including those offending headers.

That seems extremely difficult to arrange, since Frameworks refer to Frameworks .. 
(if the header is in /usr/include we have a mechanism to fix things, but not for Framework headers)

**Comment by ilg-ul** [comment]
> We can do this definitely in a "Darwin vendor" branch (it is quite conventional for distributions to have patches that are not present 'upstream')

I don't mind using a different URL to download the archives for the macOS builds, as long as the URL is persistent and the naming scheme remains stable over time.

**Comment by ilg-ul** [comment]
> I cannot be certain that such a work-around will be acceptable to upstream GCC

I don't know if this helps, disregard it if you already did so, but you can make your work-around to compile conditionally only for Apple builds, such that other targets are not affected.

**Comment by ilg-ul** [comment]
There is one more GCC casualty, `git` builds but fails a minimal test: 

```console
[/Users/ilg/.local/xbb/bin/git clone https://github.com/xpack-dev-tools/content.git content.git]
Cloning into 'content.git'...
fatal: unable to access 'https://github.com/xpack-dev-tools/content.git/': The requested URL returned error: 405
```

After rebuilding with clang, it is ok. I don't have time to investigate now, I'll leave it with clang.

I'm starting to feel that GCC might not be ready for building my development environment on Apple Silicon... :-(

**Comment by iains** [comment]
the issue with --enable-fully-dynamic-string has been fixed upstream
(it is still not needed for Darwin tho)

**Comment by ilg-ul** [comment]
I already removed it from my scripts.

Thank you.
